src/main/java/com/google/devtools/build/lib/skyframe/ActionExecutionFunction.java

```diff
@@ -274,7 +275,11 @@ private SkyValue computeInternal(SkyKey skyKey, Environment env)
     }
 
     CheckInputResults checkedInputs = null;
-    NestedSet<Artifact> allInputs = state.allInputs.getAllInputs();
+    NestedSet<Artifact> allInputs =
+        state.allInputs.getAllInputs(
+            !skyframeActionExecutor
```

Comment:
My understanding is that blaze can always assume derived input headers exist (at least virtually) and can be statted. Ulf and I discussed this, but he couldn't recall exactly how it all fits together internally. (E.g., when remote `grep-includes` is used, how is metadata for the header found? The forge client must be able to find metadata by path somehow.) I assume janakr knows the full story.

Reply:
There's nothing special happening internally, and it shouldn't be different from Bazel. Blaze will explicitly build any derived input headers, and their digests are available in memory. The only difference between Bazel and Blaze here should be BwtB, where a generated (but not downloaded) header needs to be locally include scanned. Blaze would handle this via the FUSE mount for output files. How is Bazel going to handle it?

Reply:
Bazel BwoB requires having the metadata available to know if a file needs to be downloaded. As you say, derived input headers are necessarily built—or in the case of BwoB, in a remote cache—by the time dependent include scanning runs. But, without the changes in this file, derived input header metadata is not available to the include scanning action to know if the derived headers need to be downloaded.

When I was looking into this issue, I thought there would be an analogous issue with remote include scanning. Because in order to run `grep-includes` on a header, you need to have the metadata for the header to tell the remote execution system. How does the blaze execution client figure out the metadata of the inputs to this spawn? https://github.com/bazelbuild/bazel/blob/5149408a6c2500cb1701880575c41640d4fd1c2f/src/main/java/com/google/devtools/build/lib/includescanning/SpawnIncludeScanner.java#L354-L355 Is it just statting them in the FUSE output tree?